File: Claude/Python/2025/Day2b_Solution_2025.py
with open("Day2_Input.txt") as f:
    data = f.read().strip()

ranges = []
for part in data.split(','):
    part = part.strip()
    if not part:
        continue
    a, b = part.split('-')
    ranges.append((int(a), int(b)))

max_val = max(b for a, b in ranges)
max_digits = len(str(max_val))

invalid = set()
for L in range(1, max_digits // 2 + 1):
    start_prefix = 10 ** (L - 1) if L > 1 else 1
    end_prefix = 10 ** L
    max_k = max_digits // L
    for prefix in range(start_prefix, end_prefix):
        s = str(prefix)
        for k in range(2, max_k + 1):
            n = int(s * k)
            if n <= max_val:
                invalid.add(n)

invalid_sorted = sorted(invalid)

total = 0
for a, b in ranges:
    for n in invalid_sorted:
        if n > b:
            break
        if n >= a:
            total += n

print(total)

File: Claude/Rust/2025/Day2b_Solution_2025.rs
use std::fs;
use std::collections::HashSet;

fn main() {
    let content = fs::read_to_string("Day2_Input.txt").expect("Could not read input file");
    let content = content.trim();

    let ranges: Vec<(u64, u64)> = content
        .split(',')
        .filter(|s| !s.is_empty())
        .map(|range| {
            let (lo, hi) = range.split_once('-').expect("Invalid range format");
            (lo.parse().unwrap(), hi.parse().unwrap())
        })
        .collect();

    let mut invalid_ids: HashSet<u64> = HashSet::new();

    // Enumerate all numbers whose digit string is some pattern P repeated k>=2 times.
    // For total digit length `n` and pattern length `d` where d|n and d<n,
    // enumerate all d-digit patterns (no leading zeros) and build the repeated number.
    for total_len in 2u32..=10 {
        for d in 1..total_len {
            if total_len % d != 0 {
                continue;
            }
            let reps = total_len / d;
            let pat_start: u64 = if d == 1 { 1 } else { 10u64.pow(d - 1) };
            let pat_end: u64 = 10u64.pow(d);
            let shift: u64 = pat_end; // multiply by 10^d to shift left by d digits

            for pat in pat_start..pat_end {
                // Build: pat repeated `reps` times
                let mut num = pat;
                for _ in 1..reps {
                    num = num * shift + pat;
                }

                for &(lo, hi) in &ranges {
                    if num >= lo && num <= hi {
                        invalid_ids.insert(num);
                        break;
                    }
                }
            }
        }
    }

    let total: u64 = invalid_ids.iter().sum();
    println!("{}", total);
}
